mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -213,32 +221,36 @@ class StringIndexer @Since("1.4.0") (
     val labelsArray = $(stringOrderType) match {
       case StringIndexer.frequencyDesc =>
         val sortFunc = StringIndexer.getSortFunc(ascending = false)
-        countByValue(dataset, inputCols).map { counts =>
+        val orgStrings = countByValue(dataset, inputCols).toSeq
+        ThreadUtils.parmap(orgStrings, "sortingStringLabels", 8) { counts =>
           counts.toSeq.sortWith(sortFunc).map(_._1).toArray
-        }
+        }.toArray
       case StringIndexer.frequencyAsc =>
         val sortFunc = StringIndexer.getSortFunc(ascending = true)
-        countByValue(dataset, inputCols).map { counts =>
+        val orgStrings = countByValue(dataset, inputCols).toSeq
+        ThreadUtils.parmap(orgStrings, "sortingStringLabels", 8) { counts =>
           counts.toSeq.sortWith(sortFunc).map(_._1).toArray
-        }
+        }.toArray
       case StringIndexer.alphabetDesc =>
-        import dataset.sparkSession.implicits._
         dataset.persist()
-        val labels = inputCols.map { inputCol =>
-          dataset.select(inputCol).na.drop().distinct().sort(dataset(s"$inputCol").desc)
-            .as[String].collect()
-        }
+        val selectedCols = getSelectedCols(dataset, inputCols).map(collect_set(_))
+        val allLabels = dataset.select(selectedCols: _*)
+          .collect().toSeq.flatMap(_.toSeq).asInstanceOf[Seq[Seq[String]]]
```

Comment:
so this can be selecting a large number of columns and collecting it all to the driver for distinct/sort? isn't this possibly very slow if we have billions of record?

Reply:
distinct is done at executors by `collect_set` expression. Yes, sorting is done at the driver.

This is a good point. I think it depends on the cardinality of input labels. For StringIndexer, the cardinality should not be very high, as suggested billion level.

Actually, for frequency-based string order, sorting is also done at the driver, previously.

Reply:
thanks, I think this is a reasonable trade-off, with caveat then that at very high cardinality this might OOM the driver.